File: server/makc2022--dotnet/Makc2022.Layer4.Sql.Domains.DummyMain/Queries/List/Get/DomainListGetQueryInput.cs
// Copyright (c) 2022 Maxim Kuzmin. All rights reserved. Licensed under the MIT License.

using Makc2022.Layer1.Converting;
using Makc2022.Layer1.Query;
using Makc2022.Layer2.Sql.Queries.List.Get;
using Makc2022.Layer3.Sql.Sample.Mappers.EF.Entities.DummyMain;

namespace Makc2022.Layer4.Sql.Domains.DummyMain.Queries.List.Get
{
    /// <summary>
    /// Входные данные запроса на получение списка в домене.
    /// </summary>
    public class DomainListGetQueryInput : ListGetQueryInput
    {
        #region Properties

        /// <summary>
        /// Идентификаторы сущности.
        /// </summary>
        public long[]? EntityIds { get; set; }

        /// <summary>
        /// Строка идентификаторов сущности.
        /// </summary>
        public string? EntityIdsString { get; set; }

        /// <summary>
        /// Имя сущности.
        /// </summary>
        public string? EntityName { get; set; }

        /// <summary>
        /// Идентификатор сущности "DummyOneToMany".
        /// </summary>
        public long IdOfDummyOneToManyEntity { get; set; }

        /// <summary>
        /// Идентификаторы сущности "DummyOneToMany".
        /// </summary>
        public long[]? IdsOfDummyOneToManyEntity { get; set; }

        /// <summary>
        /// Строка идентификаторов сущности "DummyOneToMany".
        /// </summary>
        public string? IdsStringOfDummyOneToManyEntity { get; set; }

        /// <summary>
        /// Имя сущности "DummyOneToMany".
        /// </summary>
        public string? NameOfDummyOneToManyEntity { get; set; }

        #endregion Properties

        #region Public methods

        /// <inheritdoc/>
        public sealed override void Normalize()
        {
            base.Normalize();

            if (string.IsNullOrWhiteSpace(SortField))
            {
                SortField = nameof(MapperDummyMainEntityObject.Id);
            }

            if (string.IsNullOrWhiteSpace(SortDirection))
            {
                SortDirection = QueryOptions.SORT_DIRECTION_DESC;
            }

            if (!string.IsNullOrWhiteSpace(EntityIdsString) && (EntityIds == null || !EntityIds.Any()))
            {
                EntityIds = EntityIdsString.FromStringToNumericInt64Array();
            }

            if (!string.IsNullOrWhiteSpace(IdsStringOfDummyOneToManyEntity)
                &&
                (
                    IdsOfDummyOneToManyEntity == null
                    ||
                    !IdsOfDummyOneToManyEntity.Any()
                ))
            {
                IdsOfDummyOneToManyEntity = IdsStringOfDummyOneToManyEntity.FromStringToNumericInt64Array();
            }
        }

        #endregion Public methods
    }
}

File: server/makc2022--nodejs--nestjs/src/layer4/domains/dummy-main/queries/list/get/domain-list-get-query-input.ts
/** Copyright (c) 2022 Maxim Kuzmin. All rights reserved. Licensed under the MIT License. */

import { nameof } from '../../../../../../layer1/common/common-functions';
import { ConvertingExtension } from '../../../../../../layer1/converting/converting.extension';
import { QuerySettings } from '../../../../../../layer1/query/query-settings';
import { ListGetQueryInput } from '../../../../../../layer2/nosql-mongo/queries/list/get/item-list-query-input';
import { MapperDummyMainEntityObject } from '../../../../../../layer3/nosql-mongo/sample/mappers/typegoose/entities/dummy-main/mapper-dummy-main-entity.object';

/** Входные данные запроса на получение элемента в домене. */
export class DomainListGetQueryInput extends ListGetQueryInput {
  /** Идентификаторы сущности. */
  entityIds: string[];

  /** Строка идентификаторов сущности. */
  entityIdsString: string;

  /** Имя сущности. */
  entityName: string;

  /** Идентификатор сущности "DummyOneToMany". */
  idOfDummyOneToManyEntity: string;

  /** Идентификаторы сущности "DummyOneToMany". */
  idsOfDummyOneToManyEntity: string[];

  /** Строка идентификаторов сущности "DummyOneToMany". */
  idsStringOfDummyOneToManyEntity: string;

  /** Имя сущности "DummyOneToMany". */
  nameOfDummyOneToManyEntity: string;

  /** @inheritdoc */
  override normalize(): void {
    super.normalize();

    if (!this.sortField) {
      this.sortField = nameof<MapperDummyMainEntityObject>('id');
    }

    if (!this.sortDirection) {
      this.sortDirection = QuerySettings.SORT_DIRECTION_DESC;
    }

    if (!this.entityIdsString && (!this.entityIds || this.entityIds.length === 0)) {
      const idsExt = ConvertingExtension.extString(this.entityIdsString);

      this.entityIds = idsExt.fromStringToStringArray();
    }

    const isOk =
      this.idsStringOfDummyOneToManyEntity &&
      (!this.idsOfDummyOneToManyEntity || this.idsOfDummyOneToManyEntity.length === 0);

    if (isOk) {
      const idsExt = ConvertingExtension.extString(this.idsStringOfDummyOneToManyEntity);

      this.idsOfDummyOneToManyEntity = idsExt.fromStringToStringArray();
    }
  }
}
